src/voter/mod.rs

```diff
@@ -501,139 +495,179 @@ impl<H, N, E: Environment<H, N>, GlobalIn, GlobalOut> Voter<H, N, E, GlobalIn, G
-					// e.g. we could start the round with a base that is too
-					// high compared to other honest voters and would fail to
-					// import their votes.
-					trace!(target: "afg", "Prospective round at {} has failed with: {}.",
-						   prospective_round.round_number(),
-						   e,
+
+					let mut round = crate::round::Round::new(crate::round::RoundParams {
+						round_number: catch_up.round_number,
+						voters: self.voters.clone(),
+						base: (catch_up.base_hash, catch_up.base_number),
+					});
+
+					// import prevotes first.
+					for crate::SignedPrevote { prevote, id, signature } in catch_up.prevotes {
+						match round.import_prevote(&*self.env, prevote, id, signature) {
+							Ok(_) => {},
+							Err(e) => {
+								trace!(target: "afg",
+									"Ignoring invalid catch up, error importing prevote: {:?}",
+									e,
+								);
+
+								process_catch_up_outcome.run(CatchUpProcessingOutcome::Bad(BadCatchUp::new()));
+								return Ok(());
+							},
+						}
+					}
+
+					// then precommits.
+					for crate::SignedPrecommit { precommit, id, signature } in catch_up.precommits {
+						match round.import_precommit(&*self.env, precommit, id, signature) {
+							Ok(_) => {},
+							Err(e) => {
+								trace!(target: "afg",
+									"Ignoring invalid catch up, error importing precommit: {:?}",
+									e,
+								);
+
+								process_catch_up_outcome.run(CatchUpProcessingOutcome::Bad(BadCatchUp::new()));
+								return Ok(());
+							},
+						}
+					}
+
+					let state = round.state();
+					if !state.completable {
+						process_catch_up_outcome.run(CatchUpProcessingOutcome::Bad(BadCatchUp::new()));
+						return Ok(())
+					}
+
+					// beyond this point, we set this round to the past and
+					// start voting in the next round.
+					let mut just_completed = VotingRound::completed(
+						round,
+						self.best_round.finalized_sender(),
+						self.env.clone(),
 					);
-				}
+
+					let new_best = VotingRound::new(
+						catch_up.round_number + 1,
+						self.voters.clone(),
+						self.last_finalized_in_rounds.clone(),
+						Some(just_completed.bridge_state()),
+						self.best_round.finalized_sender(),
+						self.env.clone(),
+					);
+
+					// update last-finalized in rounds _after_ starting new round.
+					// otherwise the base could be too eagerly set forward.
+					if let Some((f_hash, f_num)) = state.finalized.clone() {
+						if f_num > self.last_finalized_in_rounds.1 {
+							self.last_finalized_in_rounds = (f_hash, f_num);
+						}
+					}
+
+					self.env.completed(
+						just_completed.round_number(),
+						just_completed.round_state(),
+						just_completed.dag_base(),
+						just_completed.historical_votes(),
+					)?;
+
+					self.past_rounds.push(&*self.env, just_completed);
+
+					// stop voting in the best round before we background it, just in case it contradicts what
+					// we're doing now.
+					self.best_round.stop_voting();
```

Comment:
If the catch-up is to the round r+1 where `self.best_round` is at `r`, then it might be dangerous to stop voting. I think any contradiction would have to arise from others misbehaving, right?